support/src/indirect_intrusive_linear_heap.h

```diff
@@ -0,0 +1,298 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+/*
+ * Copyright (C) 2016 Red Hat Inc.
+ */
+
+
+#pragma once
+
+#include "indirect_intrusive_base.h"
+
+
+namespace crimson {
+
+
+  template<typename I, typename T, IndIntruHeapData T::*heap_info, typename CmpR, typename CmpP, typename CmpL>
+  class IndIntruLinHeap : public IndIntruBase <I, T, heap_info, CmpR, CmpP, CmpL>  {
+
+    static_assert(
+      std::is_same<T,typename std::pointer_traits<I>::element_type>::value,
+      "class I must resolve to class T by indirection (pointer dereference)");
+
+    static_assert(
+      std::is_same< bool,
+      typename std::result_of<CmpR(const T&, const T&)>::type >::value,
+      "class CmpR must define operator() to take two const T& and return a bool");
+
+    static_assert(
+      std::is_same< bool,
+      typename std::result_of<CmpP(const T&, const T&)>::type >::value,
+      "class CmpP must define operator() to take two const T& and return a bool");
+
+    static_assert(
+      std::is_same< bool,
+      typename std::result_of<CmpL(const T&, const T&)>::type >::value,
+      "class CmpL must define operator() to take two const T& and return a bool");
+
+
+    class ConstIterator {
+      friend IndIntruLinHeap < I, T, heap_info, CmpR, CmpP, CmpL>;
+
+      const IndIntruLinHeap < I, T, heap_info, CmpR, CmpP, CmpL>& heap;
+      size_t                                                      index;
+
+      ConstIterator(const IndIntruLinHeap<I, T, heap_info, CmpR, CmpP, CmpL>& _heap, size_t _index) :
+	heap(_heap),
+	index(_index)
+      {
+	// empty
+      }
+
+    public:
+
+      ConstIterator(ConstIterator&& other) :
+	heap(other.heap),
+	index(other.index)
+      {
+	// empty
+      }
+
+      ConstIterator(const ConstIterator& other) :
+	heap(other.heap),
+	index(other.index)
+      {
+	// empty
+      }
+
+      ConstIterator& operator=(ConstIterator&& other) {
+	std::swap(heap, other.heap);
+	std::swap(index, other.index);
+	return *this;
+      }
+
+      ConstIterator& operator=(const ConstIterator& other) {
+	heap = other.heap;
+	index = other.index;
+      }
+
+      ConstIterator& operator++() {
+	if (index <= heap.count) {
+	  ++index;
+	}
+	return *this;
+      }
+
+      bool operator==(const ConstIterator& other) const {
+	return &heap == &other.heap && index == other.index;
+      }
+
+      bool operator!=(const ConstIterator& other) const {
+	return !(*this == other);
+      }
+
+      const T& operator*() {
+	return *heap.data[index];
+      }
+
+      const T* operator->() {
+	return &(*heap.data[index]);
+      }
+    }; // class ConstIterator
+
+
+
+    CmpR                     cmp_resv;
+    CmpP                     cmp_ready;
+    CmpL                     cmp_limit;
+
+  protected:
+    using index_t = IndIntruHeapData;
+
+    std::vector<I>           data;
+    index_t                  count;
+    T                        *resv;
```

Comment:
Are these three variables used so that identifying the next reservation/limit/ready and then popping it doesn't require two searches? How different would the code be if they weren't there? And a related comment is this is not a generic data structure -- it has a bunch of information about dmclock built into it. I don't know if that's necessarily bad, but I'm somewhat certain that the indirect intrusive heap is generic.